### ISSUE TITLE
Dart should provide a way of combining hashes

### ISSUE THREAD

**Issue by blois** [dart-lang/sdk#11617]
Basically if one overloads operator == they should also overload hashCode and generate a 'good' hash, but Dart doesn't offer any utilities to do so. This means that many users will generate hashes with poor distributions.

The JS string class internally uses http://en.wikipedia.org/wiki/Jenkins_hash_function, it would be great if there were a public API for this.

**Comment by stereotype441** [comment]
@floitschG Thanks--I failed to notice the duplicate issue when I filed #25553.

Given that this issue is "priority-unassigned" and has been open for 2.5 years, what is the likelihood of it getting addressed soon?  As I noted in #25553, an upcoming IntelliJ feature would benefit from it, so it would be nice if we could get to it soon.

It seems like it would be a pretty simple matter of renaming a class and adding some documentation.  Would it be helpful if I submitted a CL for review?

**Comment by floitschG** [comment]
It's not the actual work. Adding new classes to core-libraries have a very high bar to pass.
In general, the tendency is to remove things, not to add features. Especially, when the functionality fits nicely into a package (as is the case here).

We will discuss it, though.

**Comment by stereotype441** [comment]
Ok, thanks.  In case the context is useful to the discussion, the IntelliJ feature @jwren is working on is automatic generation of a `hashcode` getter for a class (if the user requests it).  It would be nice to be able to use JenkinsSmiHash to create the hashcode (since this would give a far higher quality hashcode than a naive approach like XORing together the hashes of the components).  But if JenkinsSmiHash is in a package, the feature would be a lot more complex, since IntelliJ would have to check for the presence of the package in pubspec.yaml and add it if it's not already present.  Also the feature wouldn't work properly in systems that don't make use of pubspec.yaml.

**Comment by bwilkerson** [comment]
And if the package containing the function needed to be added, it would have to ensure that adding it wouldn't introduce any version constraint conflicts. And then it would need to run pub in order to add the package to the .packages file. All of which would make this a very expensive operation.

If it's in a package I expect that the feature just won't use it and we'll have lower quality code being generated.

**Comment by a14n** [comment]
BTW [quiver](https://pub.dartlang.org/packages/quiver) has this kind of functions for a while (see https://github.com/google/quiver-dart/blob/master/lib/src/core/hash.dart - Commits on Nov 14, 2013). Is it worth adding quiver as dependency and using it ?

**Comment by jwren** [comment]
+1 to what @bwilkerson and @stereotype441 said: adding a dependency is a heavier weight operation than I'd like for the Dart IntelliJ Live Template support.  Adding an import isn't sufficient as pub run has the be executed.

**Comment by lrhn** [comment]
If the code already exists in the platform libraries, just being private, then exposing it might be worth considering.
We don't necessarily want to *add* things, but exposing something that's already there is less of a concern to me - even if it means adding more public names.

What I can't decide is whether it belongs more in `dart:core` or `dart:collection`. :)

That said, I'm definitely more positive towards adding things than average :)

**Comment by stereotype441** [comment]
@floitschG Have you had a chance to discuss this?

**Comment by floitschG** [comment]
Not yet. Still on my TODO list, though.

**Comment by jwren** [comment]
For the time being, IntelliJ now generates with XORs:


```
class Example {
  var a, b, c, d, e;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    return other is Example &&
        this.a == other.a &&
        this.b == other.b &&
        this.c == other.c &&
        this.d == other.d &&
        this.e == other.e;
  }

  @override
  int get hashCode {
    return a.hashCode ^ b.hashCode ^ c.hashCode ^ d.hashCode ^ e.hashCode;
  }


}
```

**Comment by bp74** [comment]
This may not add much to the discussion, but i have duplicates of the JenkinsHash class in my libraries too (StageXL, ...). I think one of the benefits of Dart over JavaScript is that i can built on top of strong core libraries, adding a feature to calculate good hash codes should be definitely in the core libraries!

**Comment by lrhn** [comment]
I'd put a hash function in package:collection.

We'd likely still have the functionality in dart:collection and dart:core, but not necessarily with as nice an interface.

**Comment by natebosch** [comment]
Any update on this? Do we think this is something we can expose from either the SDK or package:collection?

**Comment by lrhn** [comment]
A canonical way to combine hashes is planned for Dart 2.0.

**Comment by matanlurey** [comment]
Personally I think something like this would be more usable for folks:

```dart
class Object {
  static int hash(List<Object> any) {
    // ...
  }
}
```

... but even just `JenkinsHasher` is better than copy-pasting code everywhere.

**Comment by natebosch** [comment]
We could also have `jenkinsHasher.addAll(Iterable<Object>)`.

Similarly I'd be in favor of

```dart
class Object {
  static int hash(Object a, Object b, [Object c, ...]);
  static int hashAll(List<Object> objects);
}
```

**Comment by lrhn** [comment]
I'm already working on it, but I agree with the design here.

I'd love to add a proper generalized `Hasher` interface at some point, but for now we'll just use the same hasher everywhere.

**Comment by dnfield** [comment]
@lrhn any updates on this?

**Comment by rakudrama** [comment]
Looking at the difficulty in writing a `hashCode` for a 'value'-Map implementation where `==` means the same keys and values independent of insertion order, I would suggest that corelib also supply a way of generating order-independent hashes (suitable for Sets and Maps)

```dart
class Object {
  static int hash(Object a, Object b, [Object c, ...]);
  static int hashAll(List<Object> objects);
  static int hashAllUnordered(Iterable<Object> objects);
}
```

**Comment by natebosch** [comment]
> I would suggest that corelib also supply a way of generating order-independent hashes (suitable for Sets and Maps)

IMO that would be a better fit in `package:collection`.

**Comment by rakudrama** [comment]
@natebosch While hashing of collections is a natural fit for `package:collection`, I think unordered hashing has other applications (e.g. geometric invariance), and there is a discoverability tax to having the hashing methods in several places.

I did an investigation of what would be required for a general `Object.hash` to be as efficient as the hand-rolled `hashCode` methods and the methods did need to be known to the compiler (i.e. treated specially) to get there. The main problem with all of these utility methods is that they call `o.hashCode` with static type of `o` being `Object`, which is the most complex dispatch. There are patterns of specialization that would hit a sweet spot in between slow dispatch and excessive inlining, but the exact pattern of specialization depends on the compiler (dart2js would be very different to AOT).

**Comment by jmesserly** [comment]
+1 Stephen's point. In addition, package:collection can't be used from our SDK impls, which means we may have duplicate code in the SDK. Lastly, there are proposed Dart language features that may require the compilers to automatically implement hashing/equality for value types (e.g. https://github.com/dart-lang/language/issues/125).

**Comment by techfan101** [comment]
I am new to Dart and, based on this thread, wondering if giving it a chance is a mistake.

Is something this simple (and necessary) truly still an open issue after almost **seven** years of discussion?

Please tell me I'm wrong.  I am hoping an enhancement has already shipped and that someone simply forgot to close this issue.  If so, please post the new technique / method.

If the issue truly is still open, please consider promoting its priority.  Thank you.

Given the ease of implementation and testing, this sure seems like low-hanging fruit.

**Comment by dnfield** [comment]
https://dart-review.googlesource.com/c/sdk/+/73360 Is the CL, unfortunately it's gotten held up by some other langauge design considerations.

**Comment by techfan101** [comment]
@dnfield thank you for the pointer.  Its a shame something this easy is held up.  Though, in this particular case, it benefited me.  Working around its absence forced me to recognize a better design in an unrelated part of my project :)

**Comment by Erhannis** [comment]
To recap, looks like now there's a `Object.hash(...)` function, and a few similar.  Yay!
